### PR TITLE
lanraragi 0.7.8

### PR DESCRIPTION
**Pull request by carlocab**
Continuation of #76601 and #72990.

~~Merge this after #78441 to avoid introducing a merge conflict there.~~

### CODE REVIEW

**Comment by carlocab** [comment]
Now it's trying to create another file:
```
Can't load application from file "/opt/homebrew/Cellar/lanraragi/0.7.8/libexec/script/lanraragi": can't create script/minion.pid: Operation not permitted at /opt/homebrew/Cellar/lanraragi/0.7.8/libexec/script/../lib/LANraragi/Utils/Generic.pm line 139.
```
@Difegue, any chance I can persuade `lanraragi` to write this somewhere else?

**Comment by Difegue** [comment]
I logged this problem internally during the last attempt: https://github.com/Difegue/LANraragi/issues/469 but haven't gotten to implementing the change yet, so sadly atm there's no way to redirect the .pid files. :(

**Comment by carlocab** [comment]
Oh, I see. How does `lanraragi` work if I install it into a prefix that requires `sudo` to write into?

**Comment by Difegue** [comment]
If the `script` directory is write-protected, you'll indeed have to remove/circumvent the protection somehow to get the app to run at the moment.  
This problem wasn't really ever brought up by users before, and I think the `test` used to not require `sudo` ?  
It certainly worked in previous versions of the formula.   

I've moved the .pids to the temporary folder in https://github.com/Difegue/LANraragi/commit/aeb9e6dac7d4f9c402540dbdb2c5245583df2196, so even if we can't find a solution here hopefully it won't be a problem by the next upstream release. 👍

**Comment by carlocab** [comment]
Does https://github.com/Difegue/LANraragi/commit/aeb9e6dac7d4f9c402540dbdb2c5245583df2196 still write `pid` files into the installation prefix?

It seems like the `pid` files will now be written into the `public/temp` subdirectory of the install prefix, though I might just be misunderstanding.

>I think the `test` used to not require `sudo`

The `test` does not require `sudo`. I was just wondering about how you manage trying to write into directories you may not have write permissions for.

**Comment by carlocab** [comment]
Ah, wait. The patch doesn't apply cleanly...

**Comment by Difegue** [comment]
As you saw, the temporary directory can be set through environment variables, so that allows the user to redirect the files wherever he wants.  

There's been a few changes since 0.7.8 so I'm not too surprised the patch doesn't apply 😥  
Looking at the log, the modification to `Upload.pm` is inconsequential and can probably be skipped.

**Comment by carlocab** [comment]
Didn't work, sadly.
```
Error: 6-01 09:42:43.02321] [4263] [error] Can't create process id file "/private/tmp/lanraragi-test-20210601-4237-1phflpd/server.pid": Can't open file "/private/tmp/lanraragi-test-20210601-4237-1phflpd/server.pid": No such file or directory at /opt/homebrew/Cellar/lanraragi/0.7.8/bin//../libexec/lib/perl5/Mojo/Server/Prefork.pm line 42.

Can't create process id file "/private/tmp/lanraragi-test-20210601-4237-1phflpd/server.pid": Can't open file "/private/tmp/lanraragi-test-20210601-4237-1phflpd/server.pid": No such file or directory at /opt/homebrew/Cellar/lanraragi/0.7.8/bin//../libexec/lib/perl5/Mojo/Server/Prefork.pm line 42.
 is not a list of valid directories (in $self->{"directories"}) at /opt/homebrew/Cellar/lanraragi/0.7.8/bin//../libexec/lib/perl5/File/ChangeNotify/Watcher.pm line 311
     is not a list of valid directories (in $self->{"directories"})
    "__ANON__" is defined as: sub { package File::ChangeNotify::Watcher; map {-d $_;} @{$_;}; }
```

**Comment by Difegue** [comment]
Hmm, I'm kinda stumped on that one now.  
It looks like the app is picking up the custom environment variables fine and ordering the pidfile to be created in the `testpath` directory (I'm supposing that's the `/private/tmp/lanraragi-test-xxxxx` directory), so I don't get why it fails to create the file in the end.  

I'd say it's because the `testpath` directory isn't created, but normally the `lanraragi` launcher tries to `mkdir -p` the directories anyway, so that can't be it.

**Comment by carlocab** [comment]
Ok, let's try to make sure these files exist...

**Comment by Difegue** [comment]
Looks like that worked! 😮 -- I think all that's needed is to amend the test to take into account the new line for the PID location:  
```
Server PID will be at /private/tmp/lanraragi-test-20210601-59463-8505fv/server.pid
```

**Comment by carlocab** [comment]
I think that worked? The `test` still failed, but that's because `lanraragi` exited with `255` instead of `1`, and no longer returned the table flipping ASCII art. Is that expected?

**Comment by Difegue** [comment]
Oh right, there's no table flipping either and the server loads further in somehow -- Is the test launching `redis-server` as well now?

**Comment by carlocab** [comment]
The test just calls `lanraragi`, so I don't think it's launching a redis server. Not sure how it would be doing that.

**Comment by Difegue** [comment]
nvm, I'd forgotten the binary used by the homebrew formula does [launch a `redis-server` instance](https://github.com/Difegue/LANraragi/blob/c9540841fdc2437a03e487c27071744a4aa7fdfb/tools/build/homebrew/lanraragi#L55) alongside the main server, in order to provide an "all-in-one" package.  

That line used to fail during testing IIRC, so there was no redis server, leading to the table flip.  
Adding the environment variables into the test is probably what makes it "kinda" work, but it'd be easier for testing to keep the previous behavior.  

Luckily (I guess), I think we can make the line fail again by setting `LRR_DATABASE_DIRECTORY` to a wrong path on purpose. 🤠  
```
ENV["LRR_DATABASE_DIRECTORY"] = "not_actually_a_directory"
```

**Comment by carlocab** [comment]
Huh, okay. Does that mean the formula actually also builds a redis-server now? Or has it always done that?

**Comment by Difegue** [comment]
The formula just runs the `redis-server` provided in the dependencies -- I think it's always done that, yeah.

**Comment by carlocab** [comment]
> The formula just runs the `redis-server` provided in the dependencies -- I think it's always done that, yeah.

Aha, yes, sorry, I missed the `depends_on "redis"`. Ok, I've pushed your suggested change. 🤞

**Comment by carlocab** [comment]
It seems feeding the redis-server a fake directory did not give us the failure we were looking for.

**Comment by carlocab** [comment]
Also, I just realised that this looks wrong:

https://github.com/Homebrew/homebrew-core/blob/f656d2f6bff77b9a1c34b33c5f0c718e42edcbd3/Formula/lanraragi.rb#L51

Shouldn't that be something like `Formula["openssl@1.1"].opt_prefix`, so it points to the OpenSSL install prefix? Right now the variable `OPENSSL_PREFIX` is just `openssl@1.1/1.1.1g`, which isn't right at all.

**Comment by Difegue** [comment]
Indeed, it's more resilient than I thought 😥 Sorry for wasting your time.  
I'll try to figure something out, there might be need for some extra patching.  

As for the openssl thing, I admittedly can't remember; Setting the environment variable is needed to get a specific Perl package (Net::SSLeay) to build, but it'll likely work with `opt_prefix`.

File: Formula/lanraragi.rb [hunk 1]
```diff
@@ -99,6 +112,13 @@ def caveats
   end
 
   test do
+    # Make sure lanraragi writes files to a path allowed by the sandbox
+    ENV["LRR_LOG_DIRECTORY"] = ENV["LRR_TEMP_DIRECTORY"] = testpath
+    %w[server.pid shinobu.pid minion.pid].each { |file| touch file }
+
+    # Don't actually run a redis server
+    ENV["LRR_DATABASE_DIRECTORY"] = "probably_not_a_real_directory"
+
```
Comment (Difegue) [inline Formula/lanraragi.rb:121]:
```suggestion
    # Set PERL5LIB ourselves as we're not calling the launcher script
    ENV["PERL5LIB"] = libexec/"lib/perl5"
```

File: Formula/lanraragi.rb [hunk 2]
```diff
@@ -99,6 +112,13 @@ def caveats
...
     # This can't have its _user-facing_ functionality tested in the `brew test`
     # environment because it needs Redis. It fails spectacularly tho with some
     # table flip emoji. So let's use those to confirm _some_ functionality.
```
Comment (Difegue) [inline Formula/lanraragi.rb:124]:
```suggestion
    # table flip emoji. So let's use those to confirm _some_ functionality.
    # We can't call the "lanraragi" shortcut script as that'll try to kickstart a redis-server, so we have to test
    # by running "npm start" manually.
```
I can't add a suggestion to the `assert_match` line but that's where the fix lies, see the review comment. 🙂

**Review by Difegue — COMMENTED** [review]
I took a look at the homebrew tests ran on the project's CI since those still pass, and the easiest fix(that we do internally) is actually to call `npm start` directly in `assert_match` to bypass the redis-server kickstart. 😅 

```
assert_match output, shell_output("npm start --prefix #{libexec}", 61)
```
This launches the app directly, so it still tests functionality - The `lanraragi` script simply sets some environment variable defaults and starts the redis server.  
Bar some other unforeseen weirdness, this should finally make the test pass. 

I also tried applying your openssl fix and that doesn't break the build, so that can be added as well. 👍

**Comment by carlocab** [comment]
Thanks for having a look, @Difegue.

**Comment by Difegue** [comment]
Darn, we're getting closer though! I'll run a few more tests tomorrow. 😤

**Comment by Difegue** [comment]
Huh, nevermind, it only broke on ARM. 😮 Maybe it had a redis server still hanging around from a previous run?

**Comment by carlocab** [comment]
> Huh, nevermind, it only broke on ARM. 😮 Maybe it had a redis server still hanging around from a previous run?

That's possible. Let's try to re-run it...

**Comment by carlocab** [comment]
Yup, took a peek at the ARM runner and there was, in fact, a `redis-server` still running in the background. I've killed it, so, hopefully, this test will work now.

**Comment by carlocab** [comment]
Now it failed on the Intel nodes. I think there might be a `redis-server` still running on some of them. Going to go kill those now.

**Review by nandahkrishna — APPROVED** [review]
Thanks @Difegue and @carlocab! 🎉

**Pull request closed without merging**